src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -933,9 +1673,594 @@ bool W3DDisplay::Is_LetterBox_Fading()
 // 0x0073E350
 int W3DDisplay::Get_Last_Frame_Draw_Calls()
 {
-#ifdef GAME_DLL
-    return Call_Method<int, W3DDisplay>(0x0073E350, this);
-#else
+    // return Debug_Statistics::Get_Draw_Calls();
     return 0;
+}
+
+void W3DDisplay::Update_Average_FPS()
+{
+#ifdef PLATFORM_WINDOWS
+    static uint64_t lastUpdateTime64;
+    static int historyOffset;
+    static int numSamples;
+    static double fpsHistory[30];
+
+    LARGE_INTEGER PerformanceFrequency;
+    LARGE_INTEGER PerformanceCounter;
+    QueryPerformanceFrequency(&PerformanceFrequency);
+    QueryPerformanceCounter(&PerformanceCounter);
+
+#ifdef GAME_DEBUG_STRUCTS
+    if (g_theGameLogic->Get_Frame() == 15) {
+        m_performanceCounter = PerformanceCounter.QuadPart;
+    }
+#endif
+
+    double frameTime = (double)(PerformanceCounter.QuadPart - lastUpdateTime64) / (double)PerformanceFrequency.QuadPart;
+
+    if (frameTime <= 0.5f) {
+        if (historyOffset >= 30) {
+            historyOffset = 0;
+        }
+
+        fpsHistory[historyOffset] = 1.0 / frameTime;
+        historyOffset++;
+        numSamples++;
+
+        if (numSamples > 30) {
+            numSamples = 30;
+        }
+    }
+
+    if (numSamples) {
+        double average = 0.0;
+
+        for (int i1 = 0, i2 = historyOffset - 1; i1 < numSamples; i1++, i2--) {
+            if (i2 < 0) {
+                i2 = 29;
+            }
+
+            average += fpsHistory[i2];
+        }
+
+        m_averageFps = average / (double)numSamples;
+    }
+
+    lastUpdateTime64 = PerformanceCounter.QuadPart;
+#endif
+}
+
+void W3DDisplay::Gather_Debug_Stats()
+{
+#ifdef PLATFORM_WINDOWS
+    static uint64_t lastUpdateTime64;
+    static uint64_t frameTime;
+    static unsigned int frameCount;
+    static int drawCalls;
+    static int sortingPolygons;
```

Comment:
Some of these need to be zero initialized. Zero all of them for good measure.

Reply:
fixed.